From 82c150acd6f2eaea46b9f332f169e4d365dd34dd Mon Sep 17 00:00:00 2001
From: Sai Vemuri <svemuri@redhat.com>
Date: Tue, 16 Dec 2014 22:22:13 -0500
Subject: [scsi] cxgb4i: handle non-pdu-aligned rx data

Message-id: <1418768535-2637-6-git-send-email-svemuri@redhat.com>
Patchwork-id: 102110
O-Subject: [RHEL7.1 PATCH BZ 1174982 5/7] cxgb4i: handle non-pdu-aligned rx data
Bugzilla: 1174982
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

Upstream : https://git.kernel.org/cgit/linux/kernel/git/davem/net-next.git/commit/?id=f7bcd2e11107af39367fb726196f372f8ad05199

Abort the connection upon receiving of cpl_rx_data, which means the pdu cannot
be recovered from the tcp stream. This generally is due to pdu header
corruption.

Signed-off-by: Karen Xie <kxie@chelsio.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit f7bcd2e11107af39367fb726196f372f8ad05199)
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/scsi/cxgbi/cxgb4i/cxgb4i.c b/drivers/scsi/cxgbi/cxgb4i/cxgb4i.c
index 34205ca..cbb1270 100644
--- a/drivers/scsi/cxgbi/cxgb4i/cxgb4i.c
+++ b/drivers/scsi/cxgbi/cxgb4i/cxgb4i.c
@@ -1034,6 +1034,27 @@ rel_skb:
  __kfree_skb(skb);
 }
 
+static void do_rx_data(struct cxgbi_device *cdev, struct sk_buff *skb)
+{
+ struct cxgbi_sock *csk;
+ struct cpl_rx_data *cpl = (struct cpl_rx_data *)skb->data;
+ unsigned int tid = GET_TID(cpl);
+ struct cxgb4_lld_info *lldi = cxgbi_cdev_priv(cdev);
+ struct tid_info *t = lldi->tids;
+
+ csk = lookup_tid(t, tid);
+ if (!csk) {
+  pr_err("can't find connection for tid %u.\n", tid);
+ } else {
+  /* not expecting this, reset the connection. */
+  pr_err("csk 0x%p, tid %u, rcv cpl_rx_data.\n", csk, tid);
+  spin_lock_bh(&csk->lock);
+  send_abort_req(csk);
+  spin_unlock_bh(&csk->lock);
+ }
+ __kfree_skb(skb);
+}
+
 static void do_rx_iscsi_hdr(struct cxgbi_device *cdev, struct sk_buff *skb)
 {
  struct cxgbi_sock *csk;
@@ -1453,6 +1474,7 @@ cxgb4i_cplhandler_func cxgb4i_cplhandlers[NUM_CPL_CMDS] = {
  [CPL_SET_TCB_RPL] = do_set_tcb_rpl,
  [CPL_RX_DATA_DDP] = do_rx_data_ddp,
  [CPL_RX_ISCSI_DDP] = do_rx_data_ddp,
+ [CPL_RX_DATA] = do_rx_data,
 };
 
 int cxgb4i_ofld_init(struct cxgbi_device *cdev)
-- 
1.7.1